src/coreclr/jit/lowerxarch.cpp

```diff
@@ -10628,14 +10628,53 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                                         // TODO-AVX512-CQ: Ensure we can support embedded operations on RMW intrinsics
                                         isEmbeddedMask = false;
                                     }
+                                    else
+                                    {
+                                        uint32_t maskSize = genTypeSize(simdBaseType);
+                                        uint32_t operSize = genTypeSize(op2->AsHWIntrinsic()->GetSimdBaseType());
+
+                                        if (maskSize != operSize)
+                                        {
+                                            isEmbeddedMask = false;
+                                        }
+                                        else
+                                        {
+                                            // Special cases:
+                                            // 1. vinserti128 => vinserti32x4 assumes a mask size of 512/32 = 16 bits
+                                            // (type int)
+                                            // 2. vinserti32x8 assumes a mask size of 512/32 = 16 bits (type int)
+                                            // 3. vinserti64x2 assumes a mask size of 512/64 = 8 bits (type long)
+
+                                            NamedIntrinsic op2IntrinsicId  = op2->AsHWIntrinsic()->GetHWIntrinsicId();
+                                            var_types      op2SimdBaseType = op2->AsHWIntrinsic()->GetSimdBaseType();
+                                            instruction    ins =
+                                                HWIntrinsicInfo::lookupIns(op2IntrinsicId, op2SimdBaseType);
+                                            assert(ins != INS_invalid);
+                                            switch (ins)
+                                            {
+                                                case INS_vinserti128:
+                                                case INS_vinserti32x8:
+                                                    if (operSize != 4)
```

Comment:
I wouldn't be surprised if we had more cases where SimdBaseType ends up not matching the actual instruction input size.

Anything that supports EVEX encoding will have its input size defined in the instruction attributes, so it would probably be better to resolve it from there.  I added a helper recently for this:  `CodeGenInterface::instInputSize`.